File: BackEnd/PHP/laravelshop/app/Modules/Spec/Services/SpecService.php
<?php

namespace App\Modules\Spec\Services;

use App\Common\Exceptions\BusinessException;
use App\Models\Product;
use App\Models\ProductSpec;
use App\Models\SpecDefinition;
use App\Models\User;
use Symfony\Component\HttpFoundation\Response;

class SpecService
{
    public function listDefinitions(array $filters): array
    {
        return SpecDefinition::query()
            ->when(! empty($filters['categoryId']), fn ($query) => $query->where('category_id', $filters['categoryId']))
            ->orderBy('sort_order')
            ->orderBy('id')
            ->get()
            ->map(fn (SpecDefinition $definition): array => $this->serializeDefinition($definition))
            ->values()
            ->all();
    }

    public function createDefinition(User $actor, array $payload): array
    {
        $this->assertAdmin($actor);

        $definition = SpecDefinition::query()->create([
            'category_id' => $payload['categoryId'],
            'name' => $payload['name'],
            'type' => $payload['type'] ?? 'TEXT',
            'options' => $payload['options'] ?? null,
            'unit' => $payload['unit'] ?? null,
            'sort_order' => (int) ($payload['sortOrder'] ?? 0),
        ]);

        return $this->serializeDefinition($definition);
    }

    public function updateDefinition(User $actor, int $id, array $payload): array
    {
        $this->assertAdmin($actor);

        $definition = SpecDefinition::query()->find($id);
        if (! $definition) {
            throw new BusinessException('스펙 정의를 찾을 수 없습니다.', 'SPEC_DEFINITION_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        $updates = [];
        if (array_key_exists('name', $payload)) {
            $updates['name'] = $payload['name'];
        }
        if (array_key_exists('type', $payload)) {
            $updates['type'] = $payload['type'];
        }
        if (array_key_exists('options', $payload)) {
            $updates['options'] = $payload['options'];
        }
        if (array_key_exists('unit', $payload)) {
            $updates['unit'] = $payload['unit'];
        }
        if (array_key_exists('sortOrder', $payload)) {
            $updates['sort_order'] = (int) $payload['sortOrder'];
        }

        if ($updates !== []) {
            $definition->forceFill($updates)->save();
        }

        return $this->serializeDefinition($definition->fresh());
    }

    public function deleteDefinition(User $actor, int $id): array
    {
        $this->assertAdmin($actor);

        $definition = SpecDefinition::query()->find($id);
        if (! $definition) {
            throw new BusinessException('스펙 정의를 찾을 수 없습니다.', 'SPEC_DEFINITION_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        $definition->delete();

        return ['message' => '스펙 정의가 삭제되었습니다.'];
    }

    public function getProductSpecs(int $productId): array
    {
        $product = Product::query()->find($productId);
        if (! $product) {
            throw new BusinessException('상품을 찾을 수 없습니다.', 'PRODUCT_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        return ProductSpec::query()
            ->where('product_id', $productId)
            ->orderBy('sort_order')
            ->orderBy('id')
            ->get()
            ->map(fn (ProductSpec $spec): array => $this->serializeProductSpec($spec))
            ->values()
            ->all();
    }

    public function setProductSpecs(User $actor, int $productId, array $payload): array
    {
        $this->assertAdmin($actor);

        $product = Product::query()->find($productId);
        if (! $product) {
            throw new BusinessException('상품을 찾을 수 없습니다.', 'PRODUCT_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        ProductSpec::query()->where('product_id', $productId)->delete();

        foreach ($payload['specs'] as $spec) {
            ProductSpec::query()->create([
                'product_id' => $productId,
                'spec_key' => $spec['name'],
                'spec_value' => $spec['value'],
                'sort_order' => (int) ($spec['sortOrder'] ?? 0),
            ]);
        }

        return $this->getProductSpecs($productId);
    }

    private function serializeDefinition(SpecDefinition $definition): array
    {
        return [
            'id' => $definition->id,
            'categoryId' => $definition->category_id,
            'name' => $definition->name,
            'type' => $definition->type,
            'options' => $definition->options,
            'unit' => $definition->unit,
            'sortOrder' => $definition->sort_order,
            'createdAt' => optional($definition->created_at)?->toISOString(),
            'updatedAt' => optional($definition->updated_at)?->toISOString(),
        ];
    }

    private function serializeProductSpec(ProductSpec $spec): array
    {
        return [
            'id' => $spec->id,
            'name' => $spec->spec_key,
            'value' => $spec->spec_value,
            'sortOrder' => $spec->sort_order,
        ];
    }

    private function assertAdmin(User $user): void
    {
        if ($user->role !== 'ADMIN') {
            throw new BusinessException('관리자 권한이 필요합니다.', 'FORBIDDEN', Response::HTTP_FORBIDDEN);
        }
    }
}

File: BackEnd/Java/java-spring-maven-jpa-postgresql/src/main/java/com/pbshop/java/spring/maven/jpa/postgresql/spec/SpecService.java
package com.pbshop.java.spring.maven.jpa.postgresql.spec;

import java.text.Normalizer;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.pbshop.java.spring.maven.jpa.postgresql.auth.security.AuthenticatedUserPrincipal;
import com.pbshop.java.spring.maven.jpa.postgresql.category.Category;
import com.pbshop.java.spring.maven.jpa.postgresql.category.CategoryRepository;
import com.pbshop.java.spring.maven.jpa.postgresql.common.exception.BusinessException;
import com.pbshop.java.spring.maven.jpa.postgresql.common.exception.ErrorCode;
import com.pbshop.java.spring.maven.jpa.postgresql.product.Product;
import com.pbshop.java.spring.maven.jpa.postgresql.product.ProductRepository;
import com.pbshop.java.spring.maven.jpa.postgresql.product.ProductSpec;
import com.pbshop.java.spring.maven.jpa.postgresql.product.ProductSpecRepository;
import com.pbshop.java.spring.maven.jpa.postgresql.spec.dto.SpecDtos.ProductSpecValueRequest;
import com.pbshop.java.spring.maven.jpa.postgresql.spec.dto.SpecDtos.SaveSpecDefinitionRequest;
import com.pbshop.java.spring.maven.jpa.postgresql.spec.dto.SpecDtos.SetProductSpecsRequest;

@Service
@Transactional
public class SpecService {

    private final SpecDefinitionRepository specDefinitionRepository;
    private final CategoryRepository categoryRepository;
    private final ProductRepository productRepository;
    private final ProductSpecRepository productSpecRepository;
    private final ObjectMapper objectMapper;

    public SpecService(
            SpecDefinitionRepository specDefinitionRepository,
            CategoryRepository categoryRepository,
            ProductRepository productRepository,
            ProductSpecRepository productSpecRepository,
            ObjectMapper objectMapper
    ) {
        this.specDefinitionRepository = specDefinitionRepository;
        this.categoryRepository = categoryRepository;
        this.productRepository = productRepository;
        this.productSpecRepository = productSpecRepository;
        this.objectMapper = objectMapper;
    }

    @Transactional(readOnly = true)
    public List<Map<String, Object>> getDefinitions(Long categoryId) {
        List<SpecDefinition> definitions = categoryId == null
                ? specDefinitionRepository.findAllByOrderBySortOrderAscIdAsc()
                : specDefinitionRepository.findByCategoryIdOrderBySortOrderAscIdAsc(categoryId);

        return definitions.stream().map(this::toDefinitionResponse).toList();
    }

    public Map<String, Object> createDefinition(
            AuthenticatedUserPrincipal principal,
            SaveSpecDefinitionRequest request
    ) {
        requireAdmin(principal);
        Category category = categoryRepository.findById(request.categoryId())
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "카테고리를 찾을 수 없습니다."));

        String specKey = slugify(request.key());
        if (specDefinitionRepository.existsByCategoryIdAndSpecKey(category.getId(), specKey)) {
            throw new BusinessException(ErrorCode.BAD_REQUEST, "이미 등록된 스펙 키입니다.");
        }

        SpecDefinition definition = new SpecDefinition();
        applyDefinition(definition, category, request);
        definition.setSpecKey(specKey);

        return toDefinitionResponse(specDefinitionRepository.save(definition));
    }

    public Map<String, Object> updateDefinition(
            AuthenticatedUserPrincipal principal,
            Long id,
            SaveSpecDefinitionRequest request
    ) {
        requireAdmin(principal);
        SpecDefinition definition = specDefinitionRepository.findById(id)
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "스펙 정의를 찾을 수 없습니다."));

        Category category = categoryRepository.findById(request.categoryId())
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "카테고리를 찾을 수 없습니다."));

        String specKey = slugify(request.key());
        if (specDefinitionRepository.existsByCategoryIdAndSpecKeyAndIdNot(category.getId(), specKey, id)) {
            throw new BusinessException(ErrorCode.BAD_REQUEST, "이미 등록된 스펙 키입니다.");
        }

        applyDefinition(definition, category, request);
        definition.setSpecKey(specKey);

        return toDefinitionResponse(specDefinitionRepository.save(definition));
    }

    public Map<String, Object> deleteDefinition(AuthenticatedUserPrincipal principal, Long id) {
        requireAdmin(principal);
        SpecDefinition definition = specDefinitionRepository.findById(id)
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "스펙 정의를 찾을 수 없습니다."));
        specDefinitionRepository.delete(definition);
        return Map.of("message", "스펙 정의가 삭제되었습니다.");
    }

    @Transactional(readOnly = true)
    public List<Map<String, Object>> getProductSpecs(Long productId) {
        ensureProduct(productId);
        return productSpecRepository.findByProductIdOrderBySortOrderAsc(productId).stream()
                .map(this::toProductSpecResponse)
                .toList();
    }

    public List<Map<String, Object>> setProductSpecs(
            AuthenticatedUserPrincipal principal,
            Long productId,
            SetProductSpecsRequest request
    ) {
        requireAdmin(principal);
        Product product = ensureProduct(productId);
        productSpecRepository.deleteAll(productSpecRepository.findByProductIdOrderBySortOrderAsc(productId));
        product.getSpecs().clear();

        int fallbackOrder = 0;
        for (ProductSpecValueRequest item : request.specs()) {
            ProductSpec spec = new ProductSpec();
            spec.setProduct(product);
            spec.setSpecKey(item.key());
            spec.setSpecValue(item.value());
            spec.setSortOrder(item.sortOrder() == null ? fallbackOrder : item.sortOrder());
            fallbackOrder++;
            product.getSpecs().add(spec);
            productSpecRepository.save(spec);
        }

        return productSpecRepository.findByProductIdOrderBySortOrderAsc(productId).stream()
                .map(this::toProductSpecResponse)
                .toList();
    }

    private void applyDefinition(
            SpecDefinition definition,
            Category category,
            SaveSpecDefinitionRequest request
    ) {
        definition.setCategory(category);
        definition.setName(request.name());
        definition.setInputType(request.inputType() == null || request.inputType().isBlank() ? "TEXT" : request.inputType());
        definition.setOptionsJson(writeOptionsJson(request.options()));
        definition.setUnit(request.unit());
        definition.setSortOrder(request.sortOrder() == null ? 0 : request.sortOrder());
        definition.setFilterable(request.filterable() == null || request.filterable());
    }

    private Map<String, Object> toDefinitionResponse(SpecDefinition definition) {
        Map<String, Object> response = new LinkedHashMap<>();
        response.put("id", definition.getId());
        response.put("categoryId", definition.getCategory().getId());
        response.put("categoryName", definition.getCategory().getName());
        response.put("key", definition.getSpecKey());
        response.put("name", definition.getName());
        response.put("inputType", definition.getInputType());
        response.put("options", readOptionsJson(definition.getOptionsJson()));
        response.put("unit", definition.getUnit() == null ? "" : definition.getUnit());
        response.put("sortOrder", definition.getSortOrder());
        response.put("filterable", definition.isFilterable());
        return response;
    }

    private Map<String, Object> toProductSpecResponse(ProductSpec spec) {
        return Map.of(
                "id", spec.getId(),
                "key", spec.getSpecKey(),
                "value", spec.getSpecValue(),
                "sortOrder", spec.getSortOrder()
        );
    }

    private Product ensureProduct(Long productId) {
        return productRepository.findById(productId)
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "상품을 찾을 수 없습니다."));
    }

    private void requireAdmin(AuthenticatedUserPrincipal principal) {
        if (principal == null) {
            throw new BusinessException(ErrorCode.UNAUTHORIZED, "인증이 필요합니다.");
        }
        if (!"ADMIN".equalsIgnoreCase(principal.role())) {
            throw new BusinessException(ErrorCode.FORBIDDEN, "관리자 권한이 필요합니다.");
        }
    }

    private String writeOptionsJson(List<String> options) {
        try {
            return options == null || options.isEmpty() ? null : objectMapper.writeValueAsString(options);
        } catch (Exception exception) {
            throw new BusinessException(ErrorCode.BAD_REQUEST, "스펙 옵션 직렬화에 실패했습니다.");
        }
    }

    private List<String> readOptionsJson(String optionsJson) {
        try {
            return optionsJson == null || optionsJson.isBlank()
                    ? List.of()
                    : objectMapper.readValue(optionsJson, new TypeReference<List<String>>() {
                    });
        } catch (Exception exception) {
            throw new BusinessException(ErrorCode.INTERNAL_ERROR, "스펙 옵션 역직렬화에 실패했습니다.");
        }
    }

    private String slugify(String input) {
        String normalized = Normalizer.normalize(input, Normalizer.Form.NFD)
                .replaceAll("\\p{M}", "")
                .toLowerCase(Locale.ROOT)
                .replaceAll("[^a-z0-9]+", "_")
                .replaceAll("^_+|_+$", "");

        return normalized.isBlank() ? "spec" : normalized;
    }
}
